safenode/src/network/command.rs

```diff
@@ -0,0 +1,126 @@
+use futures::channel::oneshot;
+use libp2p::{multiaddr::Protocol, request_response::ResponseChannel, Multiaddr, PeerId};
+use std::collections::{hash_map, HashSet};
+use tracing::warn;
+use xor_name::XorName;
+
+/// Commands to send to the Swarm
+#[derive(Debug)]
+pub(crate) enum CmdToSwarm {
+    StartListening {
+        addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    Dial {
+        peer_id: PeerId,
+        peer_addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    StoreData {
+        xor_name: XorName,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    GetDataProviders {
+        xor_name: XorName,
+        sender: oneshot::Sender<HashSet<PeerId>>,
+    },
+    SendRequest {
+        req: Request,
+        peer: PeerId,
+        sender: oneshot::Sender<Result<Response>>,
+    },
+    SendResponse {
+        resp: Response,
+        channel: ResponseChannel<Response>,
+    },
+}
+
+impl NetworkSwarmLoop {
+    pub(crate) fn handle_command(&mut self, command: CmdToSwarm) -> Result<(), Error> {
+        match command {
+            CmdToSwarm::StartListening { addr, sender } => {
+                let _ = match self.swarm.listen_on(addr) {
+                    Ok(_) => sender.send(Ok(())),
+                    Err(e) => sender.send(Err(e.into())),
+                };
+            }
+            CmdToSwarm::Dial {
+                peer_id,
+                peer_addr,
+                sender,
+            } => {
+                if let hash_map::Entry::Vacant(e) = self.pending_dial.entry(peer_id) {
+                    let _routing_update = self
+                        .swarm
+                        .behaviour_mut()
+                        .kademlia
+                        .add_address(&peer_id, peer_addr.clone());
+                    match self
+                        .swarm
+                        .dial(peer_addr.with(Protocol::P2p(peer_id.into())))
+                    {
+                        Ok(()) => {
+                            let _ = e.insert(sender);
+                        }
+                        Err(e) => {
+                            let _ = sender.send(Err(e.into()));
+                        }
+                    }
+                } else {
+                    warn!("Already dialing peer.");
+                }
+            }
+            // todo: the `provider` api should not be used for chunks/dbcs.
+            // 1. get the closest nodes to the data
+            // 2. store data in them directly, not via provider
+            CmdToSwarm::StoreData { xor_name, sender } => {
+                let query_id = self
+                    .swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .start_providing(xor_name.0.to_vec().into())?;
+                let _ = self.pending_start_providing.insert(query_id, sender);
+            }
+            CmdToSwarm::GetDataProviders { xor_name, sender } => {
```

Comment:
Is there possibly a name that's a bit more specific we could use rather than 'data provider'? This is the node or the peer that has the piece of data being requested? Are only some nodes/peers providers?

Reply:
Sorry, `providers` are something that we should not use for chunks/dbcs, atleast that's what I gathered from David's presentation. I will work on everything related to `providers/data` in a follow-up PR.